File: crossplane-dapr/producer-springboot/src/contractTest/java/com/agnostic/producerspringboot/ProducerPactVerificationTest.java
package com.agnostic.producerspringboot;

import au.com.dius.pact.provider.PactVerifyProvider;
import au.com.dius.pact.provider.junit5.MessageTestTarget;
import au.com.dius.pact.provider.junit5.PactVerificationContext;
import au.com.dius.pact.provider.junit5.PactVerificationInvocationContextProvider;
import au.com.dius.pact.provider.junitsupport.Provider;
import au.com.dius.pact.provider.junitsupport.loader.PactFolder;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.TestTemplate;
import org.junit.jupiter.api.extension.ExtendWith;

@Provider("order-event-producer-springboot")
@PactFolder("../consumer-springboot/pacts")
@ExtendWith(PactVerificationInvocationContextProvider.class)
class ProducerPactVerificationTest {
    private final ObjectMapper objectMapper = new ObjectMapper();

    @BeforeEach
    void before(PactVerificationContext context) {
        context.setTarget(new MessageTestTarget());
    }

    @TestTemplate
    void verifyPact(PactVerificationContext context) {
        context.verifyInteraction();
    }

    @PactVerifyProvider("an order created v1 event from producer-springboot")
    String orderCreatedMessage() throws Exception {
        return objectMapper.writeValueAsString(
            Map.of(
                "id", "ORD-SB-300",
                "amount", 55.1,
                "eventVersion", "v1"));
    }
}

File: crossplane-dapr/producer/src/contractTest/kotlin/com/agnostic/producer/contract/ProducerPactVerificationTest.kt
package com.agnostic.producer.contract

import au.com.dius.pact.provider.PactVerifyProvider
import au.com.dius.pact.provider.junit5.MessageTestTarget
import au.com.dius.pact.provider.junit5.PactVerificationContext
import au.com.dius.pact.provider.junit5.PactVerificationInvocationContextProvider
import au.com.dius.pact.provider.junitsupport.Provider
import au.com.dius.pact.provider.junitsupport.loader.PactFolder
import com.agnostic.common.events.OrderCreatedV1
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.TestTemplate
import org.junit.jupiter.api.extension.ExtendWith
import java.math.BigDecimal

@Provider("order-event-producer")
@PactFolder("../consumer/pacts")
@ExtendWith(PactVerificationInvocationContextProvider::class)
class ProducerPactVerificationTest {
    private val pactJson = Json { encodeDefaults = true }

    @BeforeEach
    fun before(context: PactVerificationContext) {
        context.target = MessageTestTarget()
    }

    @TestTemplate
    fun verifyPact(context: PactVerificationContext) {
        context.verifyInteraction()
    }

    @PactVerifyProvider("an order created v1 event")
    fun orderCreatedMessage(): String {
        val event = OrderCreatedV1(id = "ORD-300", amount = BigDecimal("55.10"))
        return pactJson.encodeToString(event)
    }
}
